src/main/java/org/commcare/formplayer/session/FormSession.java

```diff
@@ -138,13 +144,14 @@ public FormSession(UserSqlSandbox sandbox,
                        @Nullable SessionFrame sessionFrame,
                        RemoteInstanceFetcher instanceFetcher) throws Exception {
 
-        this.formDef = formDef;
-        session = new SerializableFormSession(
+        // use this.formDef to mutate (e.g., inject instance content, set callout handler)
+        this.formDef = FormDefStringSerializer.deserialize(serializableFormDefinition.getSerializedFormDef());
+        this.session = new SerializableFormSession(
                 domain, appId, TableBuilder.scrubName(username), asUser, caseId,
-                postUrl, menuSessionId, formDef.getTitle(), oneQuestionPerScreen,
+                postUrl, menuSessionId, this.formDef.getTitle(), oneQuestionPerScreen,
                 locale, inPromptMode, sessionData, functionContext
         );
-        session.setFormXml(FormDefStringSerializer.serialize(this.formDef));
```

Comment:
I'm just wondering if we shouldn't keep saving the formDef to session directly for now otherwise it will make it very hard to roll this back if the need arises.

Reply:
Realized I never responded, but addressed on 0bf6f5d48060f86769fddde01d22e9a0d1367e6b